src/embind/embind.js

```diff
@@ -608,53 +611,80 @@ var LibraryEmbind = {
     '$simpleReadValueFromPointer', '$throwBindingError'],
   _embind_register_std_string: function(rawType, name) {
     name = readLatin1String(name);
+    var embindStdStringUTF8Support
+#if EMBIND_STD_STRING_UTF8_SUPPORT
+    = true;
+#else
+    = false;
+#endif
     registerType(rawType, {
         name: name,
         'fromWireType': function(value) {
             var length = HEAPU32[value >> 2];
-            var a = new Array(length);
-            for (var i = 0; i < length; ++i) {
-                a[i] = String.fromCharCode(HEAPU8[value + 4 + i]);
+
+            function mapsToStdString(name) {
+                //process only std::string bindings with UTF8 support,
+                //in contrast to e.g. std::basic_string<unsigned char>
+                return name === "std::string";
             }
+
+            var str;
+            if(embindStdStringUTF8Support && mapsToStdString(name)) {
+                str = Module['UTF8ToString'](value + 4, length);
+            }
+            else {
+                var a = new Array(length);
+                for (var i = 0; i < length; ++i) {
+                    a[i] = String.fromCharCode(HEAPU8[value + 4 + i]);
+                }
+                str = a.join('');
+            }
+
             _free(value);
-            return a.join('');
+            
+            return str;
         },
         'toWireType': function(destructors, value) {
             if (value instanceof ArrayBuffer) {
                 value = new Uint8Array(value);
             }
+            
+            var getLength;
+            var valueIsOfTypeString = (typeof value === 'string');
 
-            function getTAElement(ta, index) {
-                return ta[index];
-            }
-            function getStringElement(string, index) {
-                return string.charCodeAt(index);
+            if (!(valueIsOfTypeString || value instanceof Uint8Array || value instanceof Uint8ClampedArray || value instanceof Int8Array)) {
+                throwBindingError('Cannot pass non-string to std::string');
             }
-            var getElement;
-            if (value instanceof Uint8Array) {
-                getElement = getTAElement;
-            } else if (value instanceof Uint8ClampedArray) {
-                getElement = getTAElement;
-            } else if (value instanceof Int8Array) {
-                getElement = getTAElement;
-            } else if (typeof value === 'string') {
-                getElement = getStringElement;
+            if (valueIsOfTypeString && embindStdStringUTF8Support) {
```

Comment:
Should this check mapsToStdString too? Otherwise `std::basic_string<unsigned char>` would be interpreted as UTF-8 too.

Reply:
Good point, should be solved with the introduction of the 'enableStdStringUTF8Support' variable (see change around line 630).